proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1565 @@
+- **rule** — the combination of: 
+    - a textual description of what constitutes a valid value;
+    - a function that asserts the rule against an actual value;
+    - a violation message template, used to report invalid values.
+    - **custom rule** — a rule based on a function (either one written by a user or provided in a `ytt` module).
+    - **named rule** — a rule provided with `ytt` expressed through a keyword argument to an `@schema/validation` or `@assert/validate`.
+- **schema base document** — the document instantiated from a schema; when the schema describes Data Values, it is the annotated default data values document.
+- **validation** — the binding of one or more "rule"s to a "node".
+- **violation** — an instance of where a given value of a "node" fails a "validation".
+
+## Proposal
+
+Implement all plumbing logic required to provide high-quality data validation over Data Values and output documents.
+
+Mechanically, this means:
+
+Extend `@schema/...` to be able to annotate validations in schema:
+- introduce `@schema/validation` which attaches a validation to a type.
+- when generating default Data Values (generally, the schema base document), when a type has a validation, annotate the corresponding node with an `@assert/validate` with that validation.
+
+Extend the `@ytt:assert` module, to include the ability to validate:
+- introduce an annotation (`@assert/validate`) which attaches one or more "rules" to a "node", defining what a valid value is for that node;
+- supply a set of common assertions (functions in the `@ytt:assert` module) that covers most validation use-cases;
+- provide guidance for how to express most commonly intended validations.
+
+Further, integrate this feature into `ytt`'s processing pipeline:
+- immediately after a final Data Values is calculated for a library, automatically validate that document (and halt processing if there are any violations);
+- if the user is inspecting schema, include validations in the output;
+- immediately after the Output Document Set (i.e. the final result of evaluating a library) is produced, automatically validate each document in the set (and halt processing if there are any violations).
+
+Finally, achieve all this in a way that complements existing validation tooling:
+- be 100% compatible with Kubernetes validation facilities (i.e. OpenAPI v3 / JSON Schema validations + [Kubernetes Validation Rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)).
+
+Deliver this functionality in the following increments:
+
+1. core validation plumbing, integrated into Schema, (at least) minimum set of built-in validations.
+2. include validations in exports to OpenAPI v3 formatted Schema "inspect" / exports.
+3. address advanced use-cases by adding the ability to edit annotations in the overlay module.
+4. integrate evaluating validations in output documents.
```

Comment:
I agree: that would be the fastest path to value delivered.

I suggested this order because the primary use cases are about giving the Configuration Consumer that better (i.e. fail fast with actionable error messages) UX regarding the thing that they are directly touching: data values.
It may well be that the 4th part there is just so easy to accomplish (given that schema validation feature _depends on_ general validations) that those two are nearly inseparable.

That said, I've been playing around with the general notion of schema (e.g. think of a document annotated as `@yaml/schema` as a lower-level version of `@data/values-schema`)... and we might want to hold back committing to what an API would look like there until we have some more data points/use-cases.